Plate Set: Assay Import - Story 2: Assay Import Graphical Layouts / Formats (single measure)

cnathe opened the pull request:
#### Rationale
The next step in the Plate Set Assay Import epic is to support more data file formats then just the tall Table format. This means that for an assay import into an assay design with the "plate metadata" property enabled, we will try a second file parsing that looks for graphical/plate layout data matrices in the file that match the expected plate size/type of the plates for the selected plate set. This story only covers the single measure case. We will extend this to multiple measure support / parsing in the future.

#### Related Pull Requests
- https://github.com/LabKey/platform/pull/5214
- https://github.com/LabKey/limsModules/pull/6

#### Changes
- only applies when the assay has plate metadata prop enabled
- data file parsing based on the plate type/size defined by the plates in the selected plate set
- allows for data grids with and without a plate identifier
- if any data grids have plate identifiers, only those are processed for import
- if no data grids have plate identifiers, verify that the number of grids doesn't exceed the number of plates in the plate set
- update PlateUtils to use LinkedHashMap to keep order of data grids when parsing
- NOTE: currently only supports numeric data grid values
- NOTE: currently only supports single measure data grid parsing

cnathe committed:
AbstractAssayTsvDataHandler to attempt to parse plate formats data if no dataRows (table format) found in file
- only applies when the assay has plate metadata prop enabled
- data file parsing based on the plate type/size defined by the plates in the selected plate set
- allows for data grids with and without a plate identifier
- if any data grids have plate identifiers, only those are processed for import
- if no data grids have plate identifiers, verify that the number of grids doesn't exceed the number of plates in the plate set
- update PlateUtils to use LinkedHashMap to keep order of data grids when parsing
- NOTE: currently only supports numeric data grid values
- NOTE: currently only supports single measure data grid parsing

cnathe committed:
Merge branch 'develop' into fb_plateSetAssayImportPart2

cnathe committed:
Fix for importing assay data in subfolder for plate set in parent/home

cnathe committed:
Fix for matching the data grids to plates by identifier and size

labkey-klum commented on api/src/org/labkey/api/assay/AbstractAssayTsvDataHandler.java:
```diff
@@ -260,6 +234,61 @@ public Map<DataType, List<Map<String, Object>>> getValidationDataMap(ExpData dat
+        return dataRows;
+    }
+
+    private Integer getPlateSetValueFromRunProps(XarContext context, AssayProvider provider, ExpProtocol protocol) throws ExperimentException
```
add the nullable annotation to this method?

cnathe replied:
done

labkey-klum commented on assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java:
```diff
@@ -508,6 +514,87 @@ else if (propEntry.getValue().isBoolean())
+        String measureName = !measureProperties.isEmpty() ? measureProperties.get(0).getName() : "Value";
+
+        // get the ordered list of plates for the plate set
+        List<Plate> plates = PlateManager.get().getPlatesForPlateSet(container, plateSetId);
```
We should verify that the `PlateSet` is resolvable from the passed in `rowId`

cnathe replied:
done

labkey-klum commented on assay/src/org/labkey/assay/plate/PlateImpl.java:
```diff
@@ -684,4 +692,16 @@ public void setPlateId(String plateId)
+    }
+
+    @Override
+    public boolean isPlateSizeMatch(int rows, int cols)
```
Rather than introducing a new method, we could just retrieve the `PlateType` from the plate and compare the shape using that object. I believe `PlateType` already supports `hashCode` and `equals`. You could introduce a constructor taking row & col for convenience.

cnathe replied:
Good idea. Done

labkey-klum commented on assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java:
```diff
@@ -508,6 +514,90 @@ else if (propEntry.getValue().isBoolean())
+                    // find the plate set plate for this identifier so we can skip those that don't match the expected plate type
+                    Plate matchingPlate = plates.stream().filter(p -> p.isIdentifierMatch(entry.getKey())).findFirst().orElse(null);
+                    double[][] plateGrid = entry.getValue();
+                    if (matchingPlate != null && matchingPlate.isPlateSizeMatch(plateGrid.length, plateGrid[0].length))
```
If we think we have plate identifiers but none (or some) of them match then we just take the matching ones? Probably a question for Rob/Nick but I wonder if none match then we fall back to indexed mode for all the data?

cnathe replied:
I will ask this during feature verification with Nick and Bernie today.

cnathe committed:
CR feedback
- parsePlateGrids() to verify plateSet resolves before attempting to get plates for it
- remove plate.isPlateSizeMath() in favor of using PlateType.equals()

cnathe committed:
use getPlateType() instead of new PlateTypeBean

cnathe committed:
PlateImpl.isIdentifierMatch() test case

cnathe committed:
Update error cases for plate data parsing
- error if assay design has more than one measure in results domain
- error if only some of the plate grids parsed have plate identifiers
- error if any of the plate grid identifiers don't match a plate in the plate set
- allow plate identifier to be in upper left corner of data grid

cnathe committed:
keep a plate map in PlateMetadataImportHelper to improve perf

cnathe committed:
Merge branch 'develop' into fb_plateSetAssayImportPart2

cnathe committed:
PlateUtils.getGridAnnotation() junit test

cnathe committed:
fix for junit expected error message

labkey-chrisj commented:
FYI I've observed that some of the error messages for importing data from plates are different from what the tests expect- I can address these when they cause test failures.  Also let's chat about what to cover in selenium tests

cnathe commented:
> FYI I've observed that some of the error messages for importing data from plates are different from what the tests expect- I can address these when they cause test failures. Also let's chat about what to cover in selenium tests

@labkey-chrisj I have a biologics branch created with the fix for those PlateSetTest failures from the error message changes. I'll open a PR after I merge your latest changes to that fb. I'll also ping you to sync on selenium testing for this story.

cnathe committed:
Merge branch 'develop' into fb_plateSetAssayImportPart2

cnathe committed:
Merge branch 'develop' into fb_plateSetAssayImportPart2
